types/src/common.ts

```diff
@@ -1,20 +1,55 @@
-import { Struct, Option, Text, bool, Vec, u16, u32, u64, Null } from '@polkadot/types'
-import { BlockNumber, Moment } from '@polkadot/types/interfaces'
-import { Codec, RegistryTypes } from '@polkadot/types/types'
+import { Struct, Option, Text, bool, Vec, u16, u32, u64, Null, U8aFixed, BTreeSet, Compact } from '@polkadot/types'
+import { BlockNumber, H256, Moment } from '@polkadot/types/interfaces'
+import { Codec, Constructor, InterfaceTypes, Registry, RegistryTypes } from '@polkadot/types/types'
+import { u8aConcat, u8aToHex } from '@polkadot/util'
 // we get 'moment' because it is a dependency of @polkadot/util, via @polkadot/keyring
 import moment from 'moment'
 import { JoyStructCustom, JoyStructDecorated } from './JoyStruct'
 import { JoyEnum } from './JoyEnum'
 
 export { JoyEnum, JoyStructCustom, JoyStructDecorated }
 
+// Adds ".sort()" during BTreeSet toU8a encoding (required by the runtime)
+// FIXME: Will not cover cases where BTreeSet is part of extrinsic args metadata
+export class JoyBTreeSet<V extends Codec> extends BTreeSet<V> {
+  public static with<V extends Codec>(valType: Constructor<V> | keyof InterfaceTypes): Constructor<JoyBTreeSet<V>> {
+    return class extends JoyBTreeSet<V> {
+      constructor(registry: Registry, value?: Uint8Array | string | Set<any>) {
+        super(registry, valType, value)
+      }
+    }
+  }
+
+  public toU8a(isBare?: boolean): Uint8Array {
+    const encoded = new Array<Uint8Array>()
+
+    if (!isBare) {
+      encoded.push(Compact.encodeU8a(this.size))
+    }
+
+    Array.from(this)
+      .sort()
```

Comment:
I can think of two 'hacks' to help in constructing BTreeMap/Set

Option 1. custom RPC calls in the joystream-node
Option 2. In rust replicate the BTreeMap/Set implementations for the types in question, and compile into WASM libraries that can be used from our javascript code?

Reply:
Another workaround is for dispatchables that expect a BtreeSet<T>, they should take a Vec<T> instead and construct the BTreeSet themselves.

Reply:
Fixed a `sort()` issue and enforced `JoyBTreeSet` values to extend `UInt` (compile-time) to make sure the sorting will always be correct.
With that we can be sure the executing extrinsics like `api.tx.storageWorkingGroup.fillOpening` we can provide application ids in any order (this can be tested ie. in Pioneer's `js` console).
As I mentioned before, runtime-side it just requires that any `BTreeSet` extrinsics arguments (unless part of structs) should have their own type name (just like `ApplicationIdSet` or `CredentialSet`)

Reply:
Excellent.